python/ray/serve/_private/application_state.py

```diff
@@ -27,61 +31,103 @@ def __init__(
         self.status = ApplicationStatus.DELETING
+        for deployment in self.get_all_deployments():
+            self.delete_deployment(deployment)
+
+    def set_deploy_failed(self, message):
+        self.status = ApplicationStatus.DEPLOY_FAILED
+        self.app_msg = message
+
+    def deploy_deployment(
+        self,
+        name: str,
+        deployment_config_proto_bytes: bytes,
+        replica_config_proto_bytes: bytes,
+        route_prefix: Optional[str],
+        deployer_job_id: Union[str, bytes],
+        docs_path: Optional[str] = None,
+        is_driver_deployment: Optional[bool] = False,
+    ) -> bool:
+        """
+        Deploys a deployment in the application.
+
+        Returns: Whether the deployment is being updated.
+        """
+
+        updating = self.deployment_state_manager.deploy(
+            deployment_name=name,
+            deployment_config_proto_bytes=deployment_config_proto_bytes,
+            replica_config_proto_bytes=replica_config_proto_bytes,
+            deployer_job_id=deployer_job_id,
+            is_driver_deployment=is_driver_deployment,
+        )
+
+        if route_prefix is not None:
+            self.route_prefix = route_prefix
+            endpoint_info = EndpointInfo(route=route_prefix)
+            self.endpoint_state.update_endpoint(name, endpoint_info)
+        else:
+            self.endpoint_state.delete_endpoint(name)
 
-    def deploy(self, deployment_params: List[Dict]) -> List[str]:
+        return updating
+
+    def deploy(self, deployment_params: List[Dict]) -> List[bool]:
         """Deploy the application.
 
         Args:
             deployment_params: list of deployment parameters, including all deployment
             information.
 
-        Returns: list of deployments which should be deleted.
+        Returns: a list of booleans indicating whether each deployment is being updated
+            or not.
         """
-
-        # Filter deployments from current deployment_params
-        # that are not used in the new deployment_params
-        to_be_deployed_deployments = {params["name"] for params in deployment_params}
-        cur_deployments_to_delete = []
-        for deployment_name in self.get_all_deployments():
-            if deployment_name not in to_be_deployed_deployments:
-                cur_deployments_to_delete.append(deployment_name)
-                self.deployments_to_delete.add(deployment_name)
+        # Delete old deployments
+        self.deployments_to_delete = set(self.get_all_deployments()).difference(
+            {params["name"] for params in deployment_params}
+        )
+        for deployment_name in self.deployments_to_delete:
+            self.delete_deployment(deployment_name)
         self.deployment_params = deployment_params
 
+        # Deploy new deployments
+        updating = []
+        for deployment in deployment_params:
+            updating.append(self.deploy_deployment(**deployment))
```

Comment:
we need to handle any failures (converting the application status) happening during the `deploy_deployment` function.